### PR TITLE
Replace map with unordered_map

### PR DESCRIPTION
**Pull request by stephomi**
Context: https://github.com/elalish/manifold/issues/483
Case study: subtract a 150k verts mesh on a 1.5M verts one (with UVs).

The edge map is one of the bottleneck when doing boolean on high poly.

Total time of `GetMeshGL()`:

|       -              |      Time     |
| -------------------- | ------------- |
| std::map (current)   | 2555 ms       |
| std::unordered_map   | 1985 ms       |
| phmap::flat_hash_map | 1450 ms       |
| (without UV)         | 1000 ms       |

phmap is header-only and from https://github.com/greg7mdp/parallel-hashmap.
For now I didn't include it, but the performance is not negligeable.
Should I do it? (git submodule? just copying the include directory?)

### CODE REVIEW

**Comment by elalish** [comment]
Thanks for the perf data! What device are you testing on and which backend are you using? Would you mind adding a comparison with the hash table we already use? https://github.com/elalish/manifold/issues/483#issuecomment-1626768739

First, because I'd like to benchmark its performance anyway, and second, because it'd be nice to use the same hash table everywhere if we find a favorite.

**Comment by stephomi** [comment]
I made another [branch](https://github.com/stephomi/manifold/tree/test_map) to test.
If you want to test, you can simply check it out, everything is included.
(to switch between method, there is a `#define` at the top of `hashtable.h`).

Device is MacBookAir M1, backend is TBB but it doesn't matter since the two sections of the code I changed are both serial.
It's the same test (1.5M and 150k subtract).

| -                              | manifold.cpp  | boolean_result.cpp |
| ------------------------------ |-------------- | ------------------ |
| std::map (master)              | 670ms         | 782ms              |
| std::unordered_map             | 227ms         | 437ms              |
| phmap::flat_hash_map           | 120ms         | 153ms              |
| hashtable.h                    | 165ms         | 200ms              |
| hashtable.h + identity         | 50ms          | -                  |
| vector<anker::svector<>>       | 50ms          | 74ms               |
| (without UV)                   | 0ms           | 0ms                |

For now a simple `vector<vector>>` linear search gave the best result (hashing mesh edges).
Maybe it's possible to improve the table result by tweaking the hash, but the vector solution is likely the best one (performance and memory as well)

---
Not sure if I understand how hashtable.h is working but it's not a drop-in replacement to map/unordered_map.
I templated it (very quick-n-dirty way) to handle arbirtrary key, not just u64.
Also the hashtable is not dynamic and you need to provide a correct size at first.
Not sure what's the expected value but when I gave the exact number of entries, it crashed.
So I'm allocating twice the number of entries.

**Comment by elalish** [comment]
Indeed, given the high likelihood of degree-6 verts, I think your vector of vectors approach makes a lot of sense. I'm reading about the small vector lib, but I don't quite understand why it's an improvement over a std::vector? If std::vector gives about the same perf, let's just use that (I figure the compilers have lots of experience optimizing std::vector).

**Comment by pca006132** [comment]
I think those small vector libraries usually flatten the vector and perform some small size optimizations, so to reduce the number of pointer chasing when accessing an element (at the expense of more expensive move etc.). I thought I modified map with unordered_map a few months ago but it seems that this is reintroduced for other parts of the code, have to say that the naming of std::map is unfortunate...

It seems to me that using vector of vectors make sense in this case. I just got back home and is now dealing with jet lag, will look at it in more details tmr.

**Comment by stephomi** [comment]
> I figure the compilers have lots of experience optimizing std::vector

It's not a compiler thing but purely a malloc thing.
In my test I got ~1.5M elements, so it's at least 1.5M malloc.
It's ok if the malloc is smart enough to give region close together.

There's a slight gain with svo, but I'm fine with std::vector<std::vector though!
I mean getMeshGL went from `2555ms` -> `1173ms` just because of the extra prop (UV)!
Without the extra prop, it was `1050ms` so I won't nitpick about ~20ms.

| -                              | manifold.cpp  | boolean_result.cpp |
| ------------------------------ |-------------- | ------------------ |
| std::vector<anker::svector<>>  | 50ms          | 74ms               |
| std::vector<std::vector<>>     | 67ms          | 80ms               |
| (without UV)                   | 0ms           | 0ms                |

Fyi `boolean_result.cpp`, here are the average size on my test (only 5 malloc)
```
size == 0 (2900)
size == 1 (1464930)
size == 2 (4932)
size == 3 (1734)
size == 4 (5)
size >= 5 (0)
```

For `manifold.cpp` I can reduce it down to 40ms (using only `std::vector<int>`) because we visit only one of `prop` neighbor.
Maybe double check that this assumption is correct.

I updated the PR, with, if you want svector, just tell me and I'll update it.

**Comment by hrgdavor** [comment]
This looks significant. How moch perf gains are we looking here?

File: src/manifold/src/boolean_result.cpp [hunk 1]
```diff
@@ -499,16 +501,15 @@ void CreateProperties(Manifold::Impl &outR, const VecDH<TriRef> &refPQ,
 
     for (const int i : {0, 1, 2}) {
       const int vert = outR.halfedge_[3 * tri + i].startVert;
-      const glm::vec3 uvw = bary[3 * tri + i];
+      const glm::vec3 &uvw = bary[3 * tri + i];
 
-      auto key = std::make_tuple(PQ, -1, -1, -1);
+      glm::ivec4 key(PQ, vert, -1, -1);
       if (oldNumProp > 0) {
-        std::get<1>(key) = vert;
```
Comment (elalish) [inline src/manifold/src/boolean_result.cpp:506]: These lines are something I just changed in #480 that I think we want to keep. The fact the tests still pass implies I need to add a test. This is to keep extra properties from being created when a mesh without properties is attached a mesh with properties.

**Review by elalish — CHANGES_REQUESTED** [review]
Excellent, thank you for a significant speed up when handling properties! And also for not introducing an extra dependency. One fix and I think it's ready.

File: src/manifold/src/boolean_result.cpp (hunk 1 above)
Comment (stephomi) [reply src/manifold/src/boolean_result.cpp:506]: Ok but maybe I'll need to revert to hashmap for `manifold.cpp` as it breaks most assumptions.
I'll check.

Comment (pca006132) [reply src/manifold/src/boolean_result.cpp:506]: One way would be to use plain array for small size hashmap and hashmap for larger size.

**Comment by pca006132** [comment]
> This looks significant. How moch perf gains are we looking here?

Seems pretty large when your mesh uses UV mapping.

File: src/manifold/src/boolean_result.cpp (hunk 1 above)
Comment (stephomi) [reply src/manifold/src/boolean_result.cpp:506]: It's fine, I can still use the `std::vector<std::vector` approach.
I made the previous assumption that one prop could belong to only 1 logical vertex but it's not the case (although gets flattened in the end).

File: src/manifold/src/boolean_result.cpp [hunk 2]
```diff
@@ -481,7 +481,10 @@ void CreateProperties(Manifold::Impl &outR, const VecDH<TriRef> &refPQ,
                           inP.halfedge_.cptrD(), inQ.halfedge_.cptrD(),
                           outR.halfedge_.cptrD(), outR.precision_}));
 
-  std::map<std::tuple<int, int, int, int>, int> propIdx;
+  using Entry = std::pair<glm::ivec3, int>;
+  int idMissProp = outR.NumVert();
+  std::vector<std::vector<Entry>> propIdx(outR.NumVert() + 1);
+  outR.meshRelation_.properties.reserve(outR.NumVert() * numProp * 1.1);
```
Comment (elalish) [inline src/manifold/src/boolean_result.cpp:487]: What is `* 1.1` for?

**Review by elalish — APPROVED** [review]
Looks great, thanks!

File: src/manifold/src/boolean_result.cpp (hunk 2 above)
Comment (stephomi) [reply src/manifold/src/boolean_result.cpp]: I've removed it.

**Pull request merged**